File: apps/ios/Tests/CrossyUITests/ReactionProximityTests.swift
// The receive-haptic gate (owner spec: soft tap only when a received sticker lands on
// or adjacent to the user's active word), pinned over the demo's 5x5 geometry so the
// row-wrap guard is exercised: numeric neighbors across a row boundary are never
// adjacency.

import CrossyUI
import XCTest

final class ReactionProximityTests: XCTestCase {
    /// The DemoRoom fixture's shape: 5x5 with blocked corners 0 and 24.
    private let puzzle = GridPuzzle(
        rows: 5, cols: 5, blocks: [0, 24], circles: [], numbers: [:])

    func test_onTheActiveWordIsNear() {
        // Across through cell 6: the full second row, cells 5...9.
        let selection = GridSelection(cell: 6, isAcross: true)
        XCTAssertTrue(
            ReactionProximity.landsNearActiveWord(cell: 8, selection: selection, puzzle: puzzle))
    }

    func test_orthogonallyAdjacentToTheWordIsNear() {
        let selection = GridSelection(cell: 6, isAcross: true)
        // Cell 1 sits directly above cell 6; cell 11 directly below.
        XCTAssertTrue(
            ReactionProximity.landsNearActiveWord(cell: 1, selection: selection, puzzle: puzzle))
        XCTAssertTrue(
            ReactionProximity.landsNearActiveWord(cell: 11, selection: selection, puzzle: puzzle))
    }

    func test_twoStepsAwayIsFar() {
        let selection = GridSelection(cell: 6, isAcross: true)
        // Cell 16 is two rows below the word: seen, not felt.
        XCTAssertFalse(
            ReactionProximity.landsNearActiveWord(cell: 16, selection: selection, puzzle: puzzle))
    }

    func test_rowWrapIsNeverAdjacency() {
        // Down through cell 2: cells 2, 7, 12, 17, 22 (the center column). Cells 9
        // and 10 are numeric neighbors of column cells across a row boundary; the
        // wrap guard keeps both far.
        let selection = GridSelection(cell: 2, isAcross: false)
        XCTAssertFalse(
            ReactionProximity.landsNearActiveWord(cell: 9, selection: selection, puzzle: puzzle))
        XCTAssertFalse(
            ReactionProximity.landsNearActiveWord(cell: 10, selection: selection, puzzle: puzzle))
        // A true lateral neighbor of the column still counts.
        XCTAssertTrue(
            ReactionProximity.landsNearActiveWord(cell: 11, selection: selection, puzzle: puzzle))
    }

    func test_outOfRangeCellsAreFar() {
        let selection = GridSelection(cell: 6, isAcross: true)
        XCTAssertFalse(
            ReactionProximity.landsNearActiveWord(cell: -1, selection: selection, puzzle: puzzle))
        XCTAssertFalse(
            ReactionProximity.landsNearActiveWord(cell: 25, selection: selection, puzzle: puzzle))
    }

    func test_aBlockSelectionHasNoWordAndNothingIsNear() {
        let selection = GridSelection(cell: 0, isAcross: true)
        XCTAssertFalse(
            ReactionProximity.landsNearActiveWord(cell: 1, selection: selection, puzzle: puzzle))
    }
}

File: apps/android/ui/src/test/kotlin/crossy/ui/ReactionProximityTests.kt
// Pins the receive-haptic gate against apps/ios ReactionProximity.swift: on the word, orthogonal
// neighbors (row-wrap guarded), never diagonals, never across the board.

package crossy.ui

import org.junit.jupiter.api.Assertions.assertFalse
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Test

class ReactionProximityTests {
    // A 3x3 with a center block; the active across word through cell 0 is (0,1,2).
    private val geometry = GridGeometry(
        rows = 3,
        cols = 3,
        blocks = setOf(4),
        circles = emptySet(),
        shadedCircles = emptySet(),
        numbers = emptyMap(),
    )
    private val standing = GridSelection(0, isAcross = true)

    @Test
    fun onTheWordAndOrthogonalNeighborsFeel() {
        assertTrue(ReactionProximity.landsNearActiveWord(1, standing, geometry), "on the word")
        assertTrue(ReactionProximity.landsNearActiveWord(3, standing, geometry), "below cell 0")
        assertTrue(ReactionProximity.landsNearActiveWord(5, standing, geometry), "below cell 2")
    }

    @Test
    fun acrossTheBoardAndDiagonalsAreSeenNotFelt() {
        // 6/8 are diagonal-or-farther from every cell of (0,1,2): silent.
        assertFalse(ReactionProximity.landsNearActiveWord(6, standing, geometry))
        assertFalse(ReactionProximity.landsNearActiveWord(8, standing, geometry))
        assertFalse(ReactionProximity.landsNearActiveWord(7, standing, geometry), "two steps down is not adjacency")
    }

    @Test
    fun rowWrapIsGuardedAndOutOfRangeIsFalse() {
        // The active down word through cell 2 is (2,5,8); cell 3 begins the next row and its
        // left-neighbor arithmetic must not wrap onto cell 2.
        val down = GridSelection(2, isAcross = false)
        assertFalse(ReactionProximity.landsNearActiveWord(3, down, geometry), "row wrap is not adjacency")
        assertFalse(ReactionProximity.landsNearActiveWord(-1, standing, geometry))
        assertFalse(ReactionProximity.landsNearActiveWord(9, standing, geometry))
    }
}
